README.md
# File Logger for `Node.js` @version 0.2.5

## installation
    npm install logm
    
## usage
    var logM = require('logm');
    logM.configure('logM-config.json');
    var logger = logM.getLogger('logger');
    
    logger.log('this is a log');
    logger.debug('this is a debug');
    logger.warn('this is a warn');
    logger.error('this is a error');
    logger.trace('this is a trace');
    
## config
loggers must be defined in config file, and pass the config file to `configure` method.   
`configure` must be called before `getLogger`

**config file sample**

    {  
        "logger1": {  
            "exclude": [  
              "DEBUG"  
            ],  
            logDir": "log/",
            "filePrefix": "test"
        },
        "logger2": {
            "exclude": [
                "LOG", "WARN"
            ],
            "logDir": "log/",
            "filePrefix": "test"
        },
        "logger3": {
            "logDir": "log/",
            "filePrefix": "tesss"
        },
        "logger4": {
            "logDir": "logg/",
            "filePrefix": "test"
        }
    }
   
## notify maintainer by mail
**usage**
	
	var logger = logM.getLogger('logger');
	logger.notifyMaintainer('maintainer@website.com',
						    'subject',
						    'content');
				
**annotation**	

1. mail host is set to author Max <max@vzhibo.tv>
2. no matter how many mails with same subject in 30 min will only be sent once.
3. if u want to send a specific mail ignore annotation 2, append a `true` param calling the `notifyMaintainer` method, which means no queue for this mail.


HISTORY.md
#HISTORY

## 0.1.0 
实现了基本日志功能，包括不同Level的log，打印到控制台和文件，每日自动更换日志文件日期后缀

## 0.2.0
添加邮件防垃圾重发机制

## 0.2.1
功能集成测试完成

## 0.2.2
增加邮件通知维护者功能

## 0.2.3
增加README.md中的API文档

## 0.2.4
增加logger打印log方法对多个参数的支持

## 0.2.5
修复读取配置文件的bug

tests/logM-spec.js
/**
 * @fileOverview logM unit test
 * @author Max
 * created at Tue Aug 19 2014 15:14:02 GMT+0800 (CST)
 */

var divider = require('./test-unit-divider');

divider.start('logger console effect');
    
    var logger = require('../src/logger').create('test');
    logger.log('this is a log', 'and second argument', 'and third argument');
    logger.debug('this is a debug', 'and second argument');
    logger.warn('this is a warn', 'and second argument');
    logger.error('this is a error', 'and second argument');
    logger.trace('this is a trace', 'and second argument');

divider.end();



//divider.start('mail test');
//
//    var mail = require('../src/mail');
//    
//    mail.send('max@vzhibo.tv',
//              'test mail 1',
//              '我爱刘花花');
//    mail.send('max@vzhibo.tv',
//              'test mail 1',
//              '我爱刘花花');
//    var action = function() {
//        mail.send('max@vzhibo.tv',
//              'test mail 1',
//              '我爱刘花花');
//    }
//    setTimeout(action, 2000);
//    
//    mail.send('max@vzhibo.tv',
//              'test mail 2',
//              '我草刘花花');
//
//divider.end();


//divider.start('file-appender write to file');
//
//    var fileAppender = require('../src/file-appender').create('log/', 'test');
//    
//    for(var i = 0; i < 1000; i++) {
//        fileAppender.appendToFile('this is the ' + i + ' log');
//    }
//
//    var count = 1000;
//    var action = function() {
//        fileAppender.appendToFile('this is the ' + count + ' log');
//        count++;
//    }
//    setInterval(action, 1000);
//
//divider.end();

//divider.start('logger notifyMaintainer');
//
//    var logger = require('../src/logger').create('test');
//    logger.notifyMaintainer('max@vzhibo.tv', 'this is a test mail', 'momoda');
//    logger.notifyMaintainer('max@vzhibo.tv', 'this is a test mail', 'momoda');
//    logger.notifyMaintainer('max@vzhibo.tv', 'this is a test mail', 'momoda', true);
//
//divider.end();
























tests/logM-integ.js
/**
 * @fileOverview logM integration test
 * @author Max
 * created at Thu Aug 21 2014 10:59:25 GMT+0800 (CST)
 */

var expect = require('expect.js');

describe('logM integration test', function() {
    var logM = require('../src/index');
    logM.configure('logM-config.json');
    
    var TIMEOUT = 31 * 1000;
    
    
    it('logger1 test single exclude', function(done) {
        this.timeout(TIMEOUT);
        var logger1 = logM.getLogger('logger1');
        logger1.log('this is log 1');
        logger1.error('this is log 2');
        logger1.warn('this is log 3');
        logger1.debug('this is log 4');
        logger1.trace('this is log 5');
        setTimeout(done, TIMEOUT - 500);
    })
    
    it('logger2 test more exclude and sam file append', function(done) {
        this.timeout(TIMEOUT);
        var logger2 = logM.getLogger('logger2');
        logger2.log('this is log 6');
        logger2.error('this is log 7');
        logger2.warn('this is log 8');
        logger2.debug('this is log 9');
        logger2.trace('this is log 10');
        setTimeout(done, TIMEOUT - 500);
    })

    it('logger3 test different file', function(done) {
        this.timeout(TIMEOUT);
        var logger3 = logM.getLogger('logger3');
        logger3.log('this is log 11');
        logger3.error('this is log 12');
        logger3.warn('this is log 13');
        logger3.debug('this is log 14');
        logger3.trace('this is log 15');
        setTimeout(done, TIMEOUT - 500);
    })
    
    it('logger4 test different dir', function(done) {
        this.timeout(TIMEOUT);
        var logger4 = logM.getLogger('logger4');
        logger4.log('this is log 16');
        logger4.error('this is log 17');
        logger4.warn('this is log 18');
        logger4.debug('this is log 19');
        logger4.trace('this is log 20');
        setTimeout(done, TIMEOUT - 500);
    })
    
    it('logger5 test logger not exists', function() {
        var loggerNotExist = logM.getLogger('logger5');
        expect(loggerNotExist).to.be('logger5 logger not found in logM-config.json');
    })
    
})    
    
    
    


    



tests/test-unit-divider.js
/**
 * @fileOverview generate a nice style console divider to divide every test unit
 * @author Max
 * created at Tue Aug 19 2014 23:24:20 GMT+0800 (CST)
 */

var colors = require('colors');
var start = function(title) {
    console.log(title.yellow);
    console.log('----------------------------------------------------------------'.grey);
}
var end = function() {
    console.log('----------------------------------------------------------------\n'.grey);
}

exports.start = start;
exports.end = end;

src/mail.js
/**
 * @fileOverview send mail method
 *               from max@vzhibo.tv
 * @author Max
 * created at Tue Aug 19 2014 11:47:31 GMT+0800 (CST)
 */

var mail = require('nodemailer');
var mailList = require('./mail-list');

var mailQueue = {};
var MAIL_INTERVAL = 30 * 60 * 1000;

var transporter = mail.createTransport({
    host: mailList.FROM_HOST,
    port: mailList.FROM_PORT,
    auth: {
        user: mailList.FROM,
        pass: mailList.FROM_PWD
    },
    secure: true
})


var send = function(obj, noQueue) {
    
    // add to queue
    if(noQueue) {
        sendMail(obj);
    } else if(mailQueue[obj.subject] == undefined) {      
        sendMail(obj);
    } else {
        var timeDiff = new Date().valueOf() - mailQueue[obj.subject].timestamp;
        if(timeDiff > MAIL_INTERVAL) {
            sendMail(obj);
        }
    }
    
    refreshMailQueueObj(obj.subject, obj);
}
// refresh mailQueue obj 
function refreshMailQueueObj(subject, obj) {
    if(mailQueue[subject] == undefined) {
        mailQueue[subject] = {};
    }
    
    mailQueue[subject].to = obj.to;
    mailQueue[subject].subject = obj.subject;
    mailQueue[subject].content = obj.content;
    mailQueue[subject].success = obj.success;
    mailQueue[subject].error = obj.error;
    mailQueue[subject].timestamp = new Date().valueOf();
}

// send mail
function sendMail(obj) {
    transporter.sendMail({
        from: mailList.FROM,
        to: obj.to,
        subject: obj.subject,
        text: obj.content
    }, function(err) {
        if(!err) {
            console.log('mail to ' + obj.to + ', subject is ' + obj.subject + ' success!');
            if(obj.success) obj.success();
        } else {
            console.log('mail to ' + obj.to + ', subject is ' + obj.subject + ' fail!');
            console.error(err);
            if(obj.error) obj.error(err);
        }
    })
}

exports.send = send;


src/index.js
/**
 * @fileOverview logM main file
 *               controller and expose interface
 * @author Max
 * created at Tue Aug 19 2014 15:23:11 GMT+0800 (CST)
 */

var fileAppender = require('./file-appender');
var loggerLevel = require('./logger-level');
var logger = require('./logger');
var fs = require('fs');

var loggers = {};

// read `logM-config.json` and generate loggers
var configure = function(configFile) {
    var configFileContent = fs.readFileSync(configFile);
    var config = JSON.parse(configFileContent);
    
    for(var i in config) {

        // construct fileAppender
        var newFileAppender = fileAppender.create(config[i].logDir, config[i].filePrefix);
        // generate exclude mask
        var mask = 0;
        if(config[i].exclude) {
            for(var j in config[i].exclude) {
                
                switch(config[i].exclude[j]) {
                    case loggerLevel.LOG:
                        mask += loggerLevel.LOG_MASK;
                        break;
                    case loggerLevel.DEBUG:
                        mask += loggerLevel.DEBUG_MASK;
                        break;
                    case loggerLevel.WARN:
                        mask += loggerLevel.WARN_MASK;
                        break;
                    case loggerLevel.ERROR:
                        mask += loggerLevel.ERROR_MASK;
                        break;
                    case loggerLevel.TRACE:
                        mask += loggerLevel.TRACE_MASK;
                        break;
                }
            }
        }

        // construct logger
        var newLogger = logger.create(i, newFileAppender, mask);
        loggers[i] = newLogger;
    }
}



var getLogger = function(category) {
    
    var logger = loggers[category];
    
    if(logger == undefined) {
        return category + ' logger not found in logM-config.json';
    } else {
        return logger;
    }
}

exports.configure = configure;
exports.getLogger = getLogger;

src/file-appender.js
/**
 * @fileOverview keep all log data
 *               write log data to file in realtime(latency accepted)
 *               name files in date format
 * @author Max
 * created at Wed Aug 20 2014 16:32:02 GMT+0800 (CST)
 */

var fs = require('fs');
var moment = require('moment');
var mail = require('./mail');
var mailList = require('./mail-list');

var FileAppender = {
    create: function(dir, prefix) {
        var obj = {};
        var logCache = '';
        var logCacheNum = 0;
        var LOG_CACHE_MAX_NUM = 100;

        var autoWriteIndex;
        var WRITE_TO_FILE_LATENCY = 30000;
        
        var today = moment();
        var LOG_FILE_POSTFIX = '.log';

        // constructor
        autoWriteIndex = setInterval(writeToFile, WRITE_TO_FILE_LATENCY);
        var exits = fs.existsSync(dir);
        if(!exits) {
            fs.mkdirSync(dir);
        }

        obj.appendToFile = function(text) {
            logCache += (text + '\n');
            logCacheNum++;
            checkLogCacheNum();
        }

        function writeToFile() {
            if(logCacheNum == 0) return;
                
            midnight();
            var filename = dir + prefix + '-' + today.format('YYYY-MM-DD') + LOG_FILE_POSTFIX;
            console.log(filename);
            fs.appendFile(filename, logCache, function(err) {
                if(err) {
                    var obj = {
                        to: mailList.MAINTAINER,
                        subject: 'LogM System Error',
                        content: 'file-appender.writeToFile > fs.appendFile error\n' +
                                 filename + '\n' +
                                 logCache
                    };
                    mail.send(obj);
                    console.trace('write to file error');
                } else {
                    console.log('write to file success');
                }
            })
            
            logCache = '';
            logCacheNum = 0;

        }
        
        // check if midnight come
        function midnight() {
            var now = moment();
            if(today.isBefore(now, 'day')) {
                console.log('midnight comes');
                today = now;
            }
        }

        // check if logCacheNum reaches max
        function checkLogCacheNum() {
            if(logCacheNum >= LOG_CACHE_MAX_NUM) {
                clearInterval(autoWriteIndex);

                writeToFile();

                autoWriteIndex = setInterval(writeToFile, WRITE_TO_FILE_LATENCY);
            }
        }
        
        return obj;
    }
}

exports.create = FileAppender.create;


src/logger.js
/**
 * @fileOverview Logger object
 *               log methods for user
 * @author Max
 * created at Tue Aug 19 2014 15:24:22 GMT+0800 (CST)
 */

var colors = require('colors');
var moment = require('moment');
var loggerLevel = require('./logger-level');
var mail = require('./mail');

colors.setTheme({
    log: 'grey',
    debug: 'green',
    warn: 'yellow',
    error: 'red',
    trace: 'blue'
});

var Logger = {
    create: function(category, fileAppender, mask) {
        var obj = {};
        
        obj.log = function() {
            if(mask & loggerLevel.LOG_MASK) return;
            var text = '';
            for(var i in arguments) {
                text += arguments[i] + ' ';
            }
            
            console.log(getPrefix(loggerLevel.LOG).log + text);
            if(fileAppender) fileAppender.appendToFile(getPrefix(loggerLevel.LOG) + text);
        }
        obj.debug = function() {
            if(mask & loggerLevel.DEBUG_MASK) return;
            var text = '';
            for(var i in arguments) {
                text += arguments[i] + ' ';
            }
            
            console.log(getPrefix(loggerLevel.DEBUG).debug + text);
            if(fileAppender) fileAppender.appendToFile(getPrefix(loggerLevel.DEBUG) + text);
        }
        obj.warn = function() {
            if(mask & loggerLevel.WARN_MASK) return;
            var text = '';
            for(var i in arguments) {
                text += arguments[i] + ' ';
            }
            
            console.warn(getPrefix(loggerLevel.WARN).warn + text);
            if(fileAppender) fileAppender.appendToFile(getPrefix(loggerLevel.WARN) + text);
        }
        obj.error = function() {
            if(mask & loggerLevel.ERROR_MASK) return;
            var text = '';
            for(var i in arguments) {
                text += arguments[i] + ' ';
            }
            
            console.error(getPrefix(loggerLevel.ERROR).error + text);
            if(fileAppender) fileAppender.appendToFile(getPrefix(loggerLevel.ERROR) + text);
        }
        obj.trace = function() {
            if(mask & loggerLevel.TRACE_MASK) return;
            var text = '';
            for(var i in arguments) {
                text += arguments[i] + ' ';
            }
            
            console.trace(getPrefix(loggerLevel.TRACE).trace + text);
            if(fileAppender) fileAppender.appendToFile(getPrefix(loggerLevel.TRACE) + text);
        }
        obj.notifyMaintainer = function(to, subject, content, noQueue) {
            var mailObj = {
                to: to,
                subject: subject,
                content: content,
                error: function(err) {
                    obj.error('mail to ' + to + ', subject is ' + subject + ' fail!' + 'error info: ' + err);
                }
            };
            mail.send(mailObj, noQueue);
        }
        
        // get yyyy-MM-dd HH:mm:ss format moment
        function getMoment() {
            return moment().format('YYYY-MM-DD HH:mm:ss');
        }
        // get prefix
        function getPrefix(level) {
            return '[' + getMoment() + '] ' + category + ' - ' + level + ': '; 
        }
        
        return obj;
    }
}

exports.create = Logger.create;